pkg/scheduler/internal/queue/scheduling_queue.go

```diff
@@ -1209,22 +1217,29 @@ func (p *PriorityQueue) movePodsToActiveOrBackoffQueue(logger klog.Logger, podIn
 	}
 }
 
-// getUnschedulablePodsWithMatchingAffinityTerm returns unschedulable pods which have
-// any affinity term that matches "pod".
+// getUnschedulablePodsWithCrossNodeTerm returns unschedulable pods which either of following conditions is met:
+// - have any affinity term that matches "pod".
+// - rejected by PodTopologySpread plugin.
 // NOTE: this function assumes lock has been acquired in caller.
-func (p *PriorityQueue) getUnschedulablePodsWithMatchingAffinityTerm(logger klog.Logger, pod *v1.Pod) []*framework.QueuedPodInfo {
+func (p *PriorityQueue) getUnschedulablePodsWithCrossNodeTerm(logger klog.Logger, pod *v1.Pod) []*framework.QueuedPodInfo {
 	nsLabels := interpodaffinity.GetNamespaceLabelsSnapshot(logger, pod.Namespace, p.nsLister)
 
 	var podsToMove []*framework.QueuedPodInfo
 	for _, pInfo := range p.unschedulablePods.podInfoMap {
+		if pInfo.UnschedulablePlugins.Has(podtopologyspread.Name) {
```

Comment:
> So this only work for in-tree plugins, for out-of-tree cross-topology plugins, the problem still exists, right?

Yes, right. People have to enable QHint if they want to avoid the same problem in custom scheduler plugins.

Reply:
> We think we can still check the podAffinity here, if notMatch, don't append, agree?

I took time thinking deeply about this, and found that we should go with the current implementation.

In short: Supposing a Pod has both topology spread and pod affinity, and a Pod is rejected only by topology spread. In this case, even `pod` here doesn't match pod-affinity, it could make the Pod schedulable.

More in-depth explanation: Supposing,
- a Pod has both topology spread and pod affinity. Topology spread's maxSkew is 1 and topology key is nodename.
- node-a has 1 Pod matching topology spread, while node-b has 3 Pod.
- both nodes have a scheduled Pod matching pod affinity.

In this case, a assigned Pod, which is on node-a and matching topology spread, can make the Pod schedulable, regardless of whether this Pod matches pod affinity or not. (because node-a already has another Pod matching pod-affinity)